File: iosApp/iosApp/features/product_detail/bdui/BduiBodySection.swift
//
//  BduiBodySection.swift
//  iosApp
//

import SwiftUI
import ComposeApp

struct BduiBodySection: View {
    let state: BduiBodyState
    let onAction: BduiActionHandler

    var body: some View {
        switch onEnum(of: state) {
        case .loading:
            HStack(spacing: 8) {
                ProgressView().scaleEffect(0.8)
                Text("Loading…")
                    .font(.body)
                    .foregroundColor(FakeShopColors.onSurfaceVariant)
            }
        case .error(let errState):
            HStack(spacing: 8) {
                Image(systemName: "exclamationmark.triangle")
                    .foregroundColor(FakeShopColors.error)
                Text(errorMessage(errState.error))
                    .font(.body)
                    .foregroundColor(FakeShopColors.onErrorContainer)
            }
            .padding(.vertical, 8)
            .padding(.horizontal, 12)
            .background(FakeShopColors.errorContainer)
            .cornerRadius(8)
        case .ready(let ready):
            BduiNodeView(
                node: ready.template.root,
                data: BindData(json: ready.bindData),
                onAction: onAction
            )
        }
    }

    private func errorMessage(_ error: BduiError) -> String {
        switch onEnum(of: error) {
        case .network: return "Couldn't load product details"
        case .notFound: return "Product layout not found"
        case .parseError: return "Unsupported layout"
        }
    }
}

File: composeApp/src/androidMain/kotlin/org/example/fakeshop_clients/features/product_detail/presentation/bdui/BduiBodySection.kt
package org.example.fakeshop_clients.features.product_detail.presentation.bdui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import org.example.fakeshop_clients.features.bdui.presentation.BduiError
import org.example.fakeshop_clients.features.productDetail.presentation.BduiBodyState

@Composable
fun BduiBodySection(state: BduiBodyState, modifier: Modifier = Modifier) {
    when (state) {
        is BduiBodyState.Loading -> BduiLoadingIndicator(modifier)
        is BduiBodyState.Error -> BduiErrorIndicator(state.error, modifier)
        is BduiBodyState.Ready -> BduiRenderer(state.template.root, state.bindData.json, modifier)
    }
}

@Composable
private fun BduiLoadingIndicator(modifier: Modifier = Modifier) {
    Row(
        modifier = modifier,
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        CircularProgressIndicator(modifier = Modifier.size(20.dp))
        Text(
            text = "Loading…",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
private fun BduiErrorIndicator(error: BduiError, modifier: Modifier = Modifier) {
    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.errorContainer
        )
    ) {
        Row(
            modifier = Modifier.padding(12.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = when (error) {
                    is BduiError.Network -> "Couldn't load product details"
                    BduiError.NotFound -> "Product layout not found"
                    is BduiError.ParseError -> "Unsupported layout"
                },
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onErrorContainer
            )
        }
    }
}
